### PR TITLE
OAuth

### PR DESCRIPTION
**Pull request by kolten**
TODO: Finish Google OAuth, it's configured to @doms application on Google (which is port :9292)

It's practically the same as the Spotify one though

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -50,236 +90,33 @@ func (a *App) Initialize(conn string) {
...
-	}
-
-	utils.RespondWithJSON(w, http.StatusCreated, u)
-}
+	fmt.Printf("Running on port %s", addr)
+	loggedRouter := handlers.LoggingHandler(os.Stdout, a.Router)
 
-func (a *App) signIn(w http.ResponseWriter, r *http.Request) {
-	var u models.User
+	// It just werks™
```
Comment (doms) [inline app.go]: 👍

File: app.go [hunk 2]
```diff
@@ -50,236 +90,33 @@ func (a *App) Initialize(conn string) {
...
+	a.Router.HandleFunc("/auth/login", a.SignIn).Methods(http.MethodPost)
+	a.Router.HandleFunc("/auth/refresh", a.Refresh).Methods(http.MethodPost)
 
-	u := models.User{ID: id}
-	if err := u.DestroyUser(a.DB); err != nil {
-		utils.RespondWithError(w, http.StatusInternalServerError, err.Error())
-		return
-	}
+	a.Router.HandleFunc("/login/spotify", a.spotifyLoginHandler)
+	a.Router.HandleFunc("/login", a.loginHandler)
```
Comment (doms) [inline app.go]: Will we need another route for `/login/google`?

File: auth_controllers.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+	if err := u.SignUp(a.DB); err != nil {
+		utils.RespondWithError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	userJSON := &models.UserJSON{
+		ID:        u.ID,
+		Name:      u.Name,
+		Email:     u.Email,
+		CreatedAt: u.CreatedAt,
```
Comment (doms) [inline auth_controllers.go]: Should we update this response to return the `access_tokens` or are we handling that differently now?

File: auth_controllers.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	// Finally, we set the client cookie for "token" as the JWT we just generated
+	// we also set an expiry time which is the same as the token itself
+	http.SetCookie(w, &http.Cookie{
+		Name:    "token",
+		Value:   tokenString,
+		Expires: expirationTime,
+		Path:    "/",
+	})
```
Comment (doms) [inline auth_controllers.go]: Noice

File: models/user.go [hunk 5]
```diff
@@ -13,15 +14,18 @@ import (
...
-	PasswordDigest string          `json:"password" db:"password_digest"`
-	AccessTokens   json.RawMessage `json:"-" db:"access_tokens"`
-	CreatedAt      time.Time       `json:"created_at"`
-	UpdatedAt      time.Time       `json:"updated_at"`
+	ID             int       `json:"id" db:"id"`
+	Name           string    `json:"name" db:"name"`
+	Email          string    `json:"email" db:"email"`
+	PasswordDigest string    `json:"password" db:"password_digest"`
+	AccessTokens   TokenMap  `json:"access_tokens" db:"access_tokens"`
+	CreatedAt      time.Time `json:"created_at"`
```
Comment (doms) [inline models/user.go]: This is really neat 🎉

File: models/user.go [hunk 6]
```diff
@@ -105,13 +142,20 @@ func (u *User) DestroyUser(db *sql.DB) error {
 	return err
 }
 
+// UpdateToken - takes the map of current user tokens and updates
+// TODO: error check this?
+func (u *User) UpdateToken(db *sql.DB) error {
+	_, err := db.Exec(`UPDATE users SET access_tokens = $1::jsonb;`, u.AccessTokens)
+	return err
+}
```
Comment (doms) [inline models/user.go]: Similar to the other methods in `user_controllers` that have an `db.Exec` call, we should add a method called `UpdateToken` and do something like

```go
// user_controllers.go
if err := u.UpdateToken(a.DB); err != nil {
    utils.RespondWithError(w, http.StatusInternalServerError, err.Error())
    return
}
```

File: app.go (hunk 2 above)
Comment (kolten) [reply app.go]: Yeah, would be better! Forgot to add that in.

File: auth_controllers.go (hunk 3 above)
Comment (kolten) [reply auth_controllers.go]: So there's two ways we can approach this now.

1. We do send the tokens to the client side, and make all the requests via `synthify-client` (which goes against the concept to my understanding) 
2. We keep the tokens internally, and make all the requests to the third party services when requested by the user. e.g. on logging in, from search (which _is_ the original concept I'm assuming)

File: models/user.go (hunk 6 above)
Comment (kolten) [reply models/user.go]: Ah, for the refresh?

Comment (doms) [reply models/user.go]: Yea! I saw your comment saying that there should be error checking for it, and I figured you wanted something similar to the other methods.

File: auth_controllers.go (hunk 3 above)
Comment (doms) [reply auth_controllers.go]: I actually like option 2 better, which yea, if it's the original concept, then we technically don't even need to return that struct, right? Just gotta send something to the client to show that "Yo, this person is authed, proceed ahead."

(I forgot that I initially built it similar to a public API when in reality, we want our proxy to be doing everything in the background)

File: models/user.go (hunk 6 above)
Comment (kolten) [reply models/user.go]: As stated in Slack earlier, this was already done 👍

**Comment by kolten** [comment]
Updated client as well +1

**Pull request merged**